### ISSUE TITLE
Bundler platform x64_mingw does not match Ruby 3.1.0 x64 on Windows

### ISSUE THREAD

**Issue by philr** [rubygems/rubygems#5269]
### Describe the problem as clearly as you can

The `:x64_mingw` Bundler platform option doesn't seem to match the RubyInstaller Ruby 3.1.0-1 x64 build. When specifying `:x64_mingw` as the platform for a gem, nothing is installed when running `bundle install` on Windows 10. This might be related to the `RUBY_PLATFORM` changing from x64-mingw to x64-mingw-ucrt.

### Did you try upgrading rubygems & bundler?

I'm using the latest versions of rubygems (v3.3.4) and bundler (v2.3.4).

### Post steps to reproduce the problem

With the RubyInstaller 3.1.0-1 x64 release installed and the following Gemfile:

```ruby
source "https://rubygems.org"

gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]
```

Run `bundle install`. This fails to install the tzinfo-data gem.

The `gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]` line is the default used by Rails applications, so this is causing problems for Rails users on Windows.

### Which command did you run?

```
bundle install
```

### What were you expecting to happen?

The tzinfo-data gem to be installed:

> Fetching gem metadata from https://rubygems.org/....
> Resolving dependencies...
> Using bundler 2.3.4
> Using concurrent-ruby 1.1.9
> Using tzinfo 2.0.4
> Using tzinfo-data 1.2021.5
> Bundle complete! 1 Gemfile dependency, 4 gems now installed.
> Use `bundle info [gemname]` to see where a bundled gem is installed.

### What actually happened?

> Resolving dependencies...
> Using bundler 2.3.4
> Bundle complete! 1 Gemfile dependency, 1 gem now installed.
> Use `bundle info [gemname]` to see where a bundled gem is installed.

## Environment

```
Bundler       2.3.4
  Platforms   ruby, x64-mingw-ucrt
Ruby          3.1.0p0 (2021-12-25 revision fb4df44d1670e9d25aef6b235a7281199a177edb) [x64-mingw-ucrt]
  Full Path   C:/Ruby/ruby-3.1.0-1-x64/bin/ruby.exe
  Config Dir  C:/ProgramData
RubyGems      3.3.4
  Gem Home    C:/Ruby/ruby-3.1.0-1-x64/lib/ruby/gems/3.1.0
  Gem Path    C:/Users/philr/.gem/ruby/3.1.0;C:/Ruby/ruby-3.1.0-1-x64/lib/ruby/gems/3.1.0
  User Home   C:/Users/philr
  User Path   C:/Users/philr/.gem/ruby/3.1.0
  Bin Dir     C:/Ruby/ruby-3.1.0-1-x64/bin
Tools
  Git         2.34.1.windows.1
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2021-12-29
Git SHA           2296a0d6cc
Released Version  true
```

### Gemfile

```ruby
source "https://rubygems.org"

gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  x64-mingw-ucrt

DEPENDENCIES
  tzinfo-data

BUNDLED WITH
   2.3.4
```

**Comment by deivid-rodriguez** [comment]
I have not a lot of context about this new ucrt platform feature, I think Windows Ruby folks like @MSP-Greg or @larskanis will have a better idea of what we need to change to support this.

**Comment by MSP-Greg** [comment]
@deivid-rodriguez @philr 

Running it locally I have:
```
[!] There was an error parsing `Gemfile`: `x64_mingw_ucrt` is not a valid platform. The available options are: <array of platforms>
```
which comes from the following, which doesn't contain `x64_mingw_ucrt`:

https://github.com/rubygems/rubygems/blob/cf3b309b2f7d67a79b11f21dca0f447f7e2f7e7b/bundler/lib/bundler/dependency.rb#L12

https://github.com/rubygems/rubygems/pull/4925 was done, and ruby/ruby CI has run with both mingw & ucrt builds for a while, but I guess this slipped through.  Not sure what the value for the hash key should be.  Certainly willing to help.

Note that Nokogiri already has `1.13.0-x64-mingw-ucrt` and `1.13.0-x64-mingw32` gems published...

**Comment by MSP-Greg** [comment]
@deivid-rodriguez

The MSYS2 system is comprised of build tools, a bash shell, and quite a few packages.  It has been used to build Windows Ruby since 2.4.

When MSYS2 was first released, all the build tools linked to what is referred to as 'msvcrt', which is short for Microsoft Visual C Runtime.  The build toolsets used for Ruby were the MinGW32 (32 bit, not widely used anymore) and MinGW64 (64 bit) tools.

msvcrt is still supported by MSFT, but most new code has been built with a newer runtime, often referred to as 'universal', hence the name ucrt.  Along with the runtime, there are also many newer system dlls/libraries.  So, MSYS2 recently released several new build toolsets, and UCRT64 was one of them.  So, Windows Ruby 3.1.0 was built with the new UCRT64 toolset.

And, re GitHub Actions, work is almost done updating it to work with Windows Ruby 3.1.0 and also the new Actions Windows image, windows-2022.  It will work now, but extension gems/repos still need some code updates... HTH

**Comment by larskanis** [comment]
I noticed the specification difference described in [this issue](https://bugs.ruby-lang.org/issues/17845#note-18) on bugs.ruby-lang.org:
> There is only one issue with bundler: x64-mingw-ucrt is treated as platform "ruby" and "mri" in the Gemfile. This is probably conform to the understanding of most users, but it is against the specification of [bundlers platform argument](https://github.com/rubygems/rubygems/blob/96e5cff3df491c4d943186804c6d03b57fcb459b/bundler/lib/bundler/man/gemfile.5.ronn#platforms) which classifies "mri" as NOT Windows.

I already tried to add the `x64-mingw-ucrt` platform to the scope of `platform: x64_mingw` in the Gemfile, but the platform code in bundler is ... ugly. So I wasn't successful. It is certainly not very hard to add `x64_mingw_ucrt` to the PLATFORM_MAP and some other places and establish this as a new Gemfile platform, but is it desired to establish a new platform there?

To clarify: rubygems and bundler already properly work with published gems for the platform `x64-mingw-ucrt`. So selection and installation of `nokogiri` or `ffi` gems works as it should (it installs the `x64-mingw-ucrt` gem preferably and falls back to the `ruby` platform). Only the `platform` key in the Gemfile isn't conforming to the specs, since `x64-mingw-ucrt` is currently in the scope of `ruby` and `mri` and not in the scope of `mingw` or `x64_mingw`.

**Comment by deivid-rodriguez** [comment]
> I already tried to add the x64-mingw-ucrt platform to the scope of platform: x64_mingw in the Gemfile, but the platform code in bundler is ... ugly. So I wasn't successful. It is certainly not very hard to add x64_mingw_ucrt to the PLATFORM_MAP and some other places and establish this as a new Gemfile platform, but is it desired to establish a new platform there?

Yeah, right, our platform code is not ideal. But think about it this way, previously we had ugly non working code, now we have ugly working much better code :)

> I already tried to add the x64-mingw-ucrt platform to the scope of platform: x64_mingw in the Gemfile, but the platform code in bundler is ... ugly. So I wasn't successful. It is certainly not very hard to add x64_mingw_ucrt to the PLATFORM_MAP and some other places and establish this as a new Gemfile platform, but is it desired to establish a new platform there?

I see, that's a nice summary of the current state of things. So, as an alternative to adding a new platform, you'd prefer that the current `x64_mingw` platform value resolved to `urct` variants, right?

**Comment by larskanis** [comment]
> So, as an alternative to adding a new platform, you'd prefer that the current `x64_mingw` platform value resolved to `urct` variants, right?

Yes. The platforms `x64-mingw32` and `x64-mingw-ucrt` are ABI incompatible - that's why we introduced a new platform string for RUBY_PLATFORM and Rubygems. But there is little to no difference between them, when it comes to enabling/disabling gems from a Gemfile.

**Comment by jume-dev** [comment]
Hello there,
this is an extremely important feature for me and it halts my migration to ruby 3.1. Any timeline for when this will be implemented? 
If not prioritized right now, maybe I can contribute a bit. Can anybody point me in a direction?

**Comment by ariccio** [comment]
@jume-dev in the meantime just add a plain `gem 'tzinfo-data'` to your gemfile.

**Comment by omokehinde** [comment]
> I have not a lot of context about this new ucrt platform feature, I think Windows Ruby folks like @MSP-Greg or @larskanis will have a better idea of what we need to change to support this.

Change gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ] to gem "tzinfo-data" and I it should work. It is similar to this issue here: https://github.com/rails/rails/issues/44952

**Comment by johnnyshields** [comment]
@deivid-rodriguez @larskanis fix is here: https://github.com/rubygems/rubygems/pull/5649
